### PR TITLE
[beta4] Possible vswhere UTF-8 fix & Differentiate between release and pre-release installations

### LINKED ISSUE #220: Failed to parse json generated by vswhere
I got following error with Windows  10(Japanese) and MSVC 2017 Community.

```
Unexpected token � in JSON at position 368: SyntaxError: Unexpected token � in JSON at position 368
    at JSON.parse (<anonymous>)
    at Object.<anonymous> (C:\Users\kkimura\.vscode\extensions\vector-of-bool.cmake-tools-0.10.2\out\src\environment.js:271:39)
    at Generator.next (<anonymous>)
    at fulfilled (C:\Users\kkimura\.vscode\extensions\vector-of-bool.cmake-tools-0.10.2\out\src\environment.js:4:58)
```
Here is output of `vswhere.exe -all -format json -product *  -legacy`.
```json
[
  {
    "instanceId": "01eb69bb",
    "installDate": "2017-07-31T06:40:09Z",
    "installationName": "VisualStudio/15.2.0+26430.16",
    "installationPath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community",
    "installationVersion": "15.0.26430.16",
    "displayName": "Visual Studio Community 2017",
    "description": "�w���A�I�[�v�� �\�[�X�A����ьX�̊J���҂̂��߂̖����Ŋ��S�ȋ@�\������� IDE",
    "enginePath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\Installer\\resources\\app\\ServiceHub\\Services\\Microsoft.VisualStudio.Setup.Service",
    "channelId": "VisualStudio.15.Release",
    "channelPath": "C:\\Users\\kkimura\\AppData\\Local\\Microsoft\\VisualStudio\\Packages\\_Channels\\4CB340F5\\catalog.json",
    "channelUri": "https://aka.ms/vs/15/release/channel",
    "releaseNotes": "https://go.microsoft.com/fwlink/?LinkId=660284#15.1.26430.16",
    "thirdPartyNotices": "https://go.microsoft.com/fwlink/?LinkId=660300"
  }
]
```

Output encoding is CP932. I think this is vswhere's bug.

### PR DESCRIPTION
**Pull request by Randshot**
## Related Issues
Fixes #220.

## Changes
This PR tries to fix the vswhere utf8 bug, where the JSON parser throws because of encoding/code page issues. The method I used is the one mentioned by heaths in the issue above.
I added a separate function that uses `execFile()` instead of `spawn()` in the hope that it fixes the issues other users encountered. Sadly, I couldn't really verify whether this fixes the problem at hand though.

The other thing that this PR does is to differentiate between release and pre-release versions of Visual Studio. For pre-release versions, the `Preview` identifier is appended to the display name.
Since I don't have the **Community** edition of VS installed, but the **Professional** one, pretty much all the `preferred-generators` tests failed. I fixed this by getting the known kits from the extension (I had to add a getter for that) and checking whether the default kit name in question is present. I redid the check for preferred generators the same way, instead of using a regex.

### CODE REVIEW

File: src/proc.ts [hunk 1]
```diff
@@ -74,6 +72,14 @@ export interface ExecutionOptions {
   cwd?: string;
 }
 
+export interface ExecFileOptions {
```
Comment (vector-of-bool) [inline src/proc.ts]: You can have this interface extend `ExecutionOptions`, just to remove duplication

File: src/proc.ts [hunk 2]
```diff
@@ -163,3 +169,94 @@ export function execute(command: string,
...
+  const final_env = util.mergeEnvironment(process.env as EnvironmentVariables, options.environment || {});
+
+  const exec_opts: proc.ExecOptionsWithStringEncoding
+      = {env: final_env, shell: options.shell, encoding: options.encoding || 'utf8'};
+
+  if (options && options.cwd) {
+    exec_opts.cwd = options.cwd;
+  }
+
+  const child: proc.ChildProcess = proc.execFile(command, args, exec_opts);
```
Comment (vector-of-bool) [inline src/proc.ts]: Use the `callback` signature of [the `child_process.execFile()` API](https://nodejs.org/api/child_process.html#child_process_child_process_execfile_file_args_options_callback). We don't need the output streaming abilities and `OutputConsumer` stuff.

File: src/kit.ts [hunk 3]
```diff
@@ -299,11 +300,13 @@ export async function vsInstallations(): Promise<VSInstallation[]> {
   const inst_ids = [] as string[];
   const vswhere_exe = path.join(thisExtensionPath(), 'res/vswhere.exe');
   const vswhere_args = ['-all', '-format', 'json', '-products', '*', '-legacy', '-prerelease'];
-  const vswhere_res = await proc.execute(vswhere_exe, vswhere_args).result;
+  const vswhere_res = await proc.execute(vswhere_exe, vswhere_args, null, {silent: true, encoding: 'utf8'}).result;
```
Comment (vector-of-bool) [inline src/kit.ts]: What happened to using `cmd.exe` and `chcp`?

Comment (Randshot) [reply src/kit.ts]: Ah, I haven't added that yet, since I am still  puzzled by the VS problem.

File: src/kit.ts [hunk 4]
```diff
@@ -299,8 +299,9 @@ export async function vsInstallations(): Promise<VSInstallation[]> {
   const installs = [] as VSInstallation[];
   const inst_ids = [] as string[];
   const vswhere_exe = path.join(thisExtensionPath(), 'res/vswhere.exe');
-  const vswhere_args = ['-all', '-format', 'json', '-products', '*', '-legacy', '-prerelease'];
-  const vswhere_res = await proc.execute(vswhere_exe, vswhere_args, null, {silent: true, encoding: 'utf8'}).result;
+  const vswhere_args =
+      ['/c', 'chcp 65001 |', vswhere_exe, '-all', '-format', 'json', '-products', '*', '-legacy', '-prerelease'];
```
Comment (vector-of-bool) [inline src/kit.ts]: Does this work? I'd think `child_process` would quote the arguments before calling CreateProcess and we'd end up with something like

```bash
cmd.exe /c "chcp 65001 |" vswhere.exe -all -format -json -products * -legacy -prerelease
```

That'll behave weird unless cmd.exe is doing its own goobery command-line parsing and not using `argv` from the CRT

Comment (vector-of-bool) [reply src/kit.ts]: Just tested, and can confirm that cmd.exe is doing something goobery:

```bash 
> cmd.exe /c "echo hi" echo foo
hi echo foo
```

Comment (Randshot) [reply src/kit.ts]: hm, it does work for me

**Pull request merged**